### ISSUE TITLE
Move the request encoder API to Resource

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6031]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#6031|
|Type|enhancement|
|Created|2012-09-30 08:37:45Z|
|Branch|https://github.com/twisted/twisted/tree/encoder-resource-6031-2|

The new API merged at [#11528](https://github.com/twisted/twisted/issues/11528) is on Site, but it doesn't allow to selectively enable encoders on specific path. It's particularly important given the SSL/Zlib attacks on user-generated requests (see https://bugzilla.redhat.com/show_bug.cgi?id=857051 for example).

<details><summary>Searchable metadata</summary>

```
trac-id__6031 6031
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__Twisted_12_3 Twisted-12.3
branch__branches_encoder_resource_6031_2 branches/encoder-resource-6031-2
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1348994265000000 1348994265000000
changetime__1350391549000000 1350391549000000
version__None None
owner__therve therve
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [35846]) Branching to 'encoder-resource-6031'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

Here it is!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [36174]) Branching to 'encoder-resource-6031-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

Thanks.  Just some minor doc issues:

  1. Seems like `EncodingResourceWrapper` shouldn't bother talking about the private interface that is used to implement the behavior it provides (since we don't want to encourage users to do anything with this interface).
  1. Would also be nice to see `EncodingResourceWrapper` be a little more explicit about what it does: (potentially) apply one of the encodings to the response body generated by the wrapped resource.  Perhaps also worth mentioning encodings are not inherited by children of the resource.
  1. in the `Resource` docstring, instead of "It serves 2 main purposes..." I would write "This serves 2 main purposes...".

That's all.  Please fix these and then merge (if the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/encoder-resource-6031-2) look good), thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set status to `closed`|
|-|-|

(In [36181]) Merge encoder-resource-6031-2 

Author: therve
Reviewer: exarkun
Fixes: [#6031](https://github.com/twisted/twisted/issues/6031)

Move the encoding mechanism logic in twisted.web from Site to a wrapper
Resource, so that users can more selectively apply it. In particular, add a
note in the documentation about SSL/zlib attacks on user-generated content.
